include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/inclusive_scan.hpp

```diff
@@ -0,0 +1,263 @@
+
+    std::vector<sycl::event> events;
+    auto root = devices()[0];
+    device_allocator<T> allocator(context(), root);
+    vector<T, device_allocator<T>> partial_sums(std::size_t(zipped_segments.size()), allocator);
+
+    // DPL futures must be kept alive, since in the future their destruction
+    // may trigger a synchronization.
+    using dpl_future_type = decltype(inclusive_scan_async(
+        __detail::dpl_policy(0), dr::__detail::direct_iterator(stdrng::begin(std::get<0>(*zipped_segments.begin()))),
+        dr::__detail::direct_iterator(stdrng::begin(std::get<0>(*zipped_segments.begin()))),
+        dr::__detail::direct_iterator(stdrng::begin(std::get<1>(*zipped_segments.begin()))), binary_op, init.value()));
+
+    std::vector<dpl_future_type> futures;
+
+    std::size_t segment_id = 0;
+    for (auto&& segs : zipped_segments)
+    {
+        auto&& [in_segment, out_segment] = segs;
+
+        auto&& q = __detail::queue(ranges::rank(in_segment));
+        auto&& local_policy = __detail::dpl_policy(ranges::rank(in_segment));
+
+        auto dist = stdrng::distance(in_segment);
+        assert(dist > 0);
+
+        auto first = stdrng::begin(in_segment);
+        auto last = stdrng::end(in_segment);
+        auto d_first = stdrng::begin(out_segment);
+
+        if (segment_id == 0 && init.has_value())
+        {
+            futures.push_back(inclusive_scan_async(local_policy, dr::__detail::direct_iterator(first),
+                                                   dr::__detail::direct_iterator(last),
+                                                   dr::__detail::direct_iterator(d_first), binary_op, init.value()));
+        }
+        else
+        {
+            futures.push_back(inclusive_scan_async(local_policy, dr::__detail::direct_iterator(first),
+                                                   dr::__detail::direct_iterator(last),
+                                                   dr::__detail::direct_iterator(d_first), binary_op));
+        }
+
+        auto dst_iter = ranges::local(partial_sums).data() + segment_id;
+
+        auto src_iter = ranges::local(out_segment).data();
+        stdrng::advance(src_iter, dist - 1);
+
+        sycl::event event = futures.back();
+
+        auto e = q.submit([&](auto&& h) {
+            h.depends_on(event);
+            h.single_task([=]() {
+                stdrng::range_value_t<O> value = *src_iter;
+                *dst_iter = value;
+            });
+        });
+
+        events.push_back(e);
+
+        segment_id++;
+    }
+
+    __detail::wait(events);
+    events.clear();
+    futures.clear();
+
+    auto&& local_policy = __detail::dpl_policy(0);
+
+    auto first = ranges::local(partial_sums).data();
+    auto last = first + partial_sums.size();
+
+    inclusive_scan_async(local_policy, first, last, first, binary_op).wait();
+
+    std::size_t idx = 0;
+    for (auto&& segs : zipped_segments)
+    {
+        auto&& [in_segment, out_segment] = segs;
```

Comment:
It seems like this line can be moved within the `if` scope. Or perhaps this loop can be rewritten to have a `if (idx == 0) continue;` at the top and remove the following `if` entirely.